### PR TITLE
Destination redshift dv2: Update type mapping

### PR DESCRIPTION
**Pull request by edgao**
https://airbytehq-team.slack.com/archives/C0690DFUDGS/p1702311247948029?thread_ts=1702299183.996059&cid=C0690DFUDGS

* jdbc sql generator: map numbers to numeric rather than float.
* redshift: override the string mapping to be explicitly varchar. we're using the postgres dialect here, which renders `SQLDataType.VARCHAR` to `VARCHAR(256)` 😓

### CODE REVIEW

File: airbyte-cdk/java/airbyte-cdk/db-destinations/src/main/java/io/airbyte/cdk/integrations/destination/jdbc/typing_deduping/JdbcSqlGenerator.java [hunk 1]
```diff
@@ -63,7 +63,7 @@ protected DataType<?> toDialectType(final AirbyteType type) {
   protected DataType<?> toDialectType(final AirbyteProtocolType airbyteProtocolType) {
     return switch (airbyteProtocolType) {
       case STRING -> SQLDataType.VARCHAR;
-      case NUMBER -> SQLDataType.FLOAT;
+      case NUMBER -> SQLDataType.NUMERIC;
```
Comment (gisripa) [inline airbyte-cdk/java/airbyte-cdk/db-destinations/src/main/java/io/airbyte/cdk/integrations/destination/jdbc/typing_deduping/JdbcSqlGenerator.java:66]: DECIMAL and NUMERIC are interchangeable in redshift context. which one seems more representative for readability ?

Comment (edgao) [reply airbyte-cdk/java/airbyte-cdk/db-destinations/src/main/java/io/airbyte/cdk/integrations/destination/jdbc/typing_deduping/JdbcSqlGenerator.java:66]: https://docs.aws.amazon.com/redshift/latest/dg/r_Numeric_types201.html#r_Numeric_types201-decimal-or-numeric-type

> In this document, decimal is the preferred term for this data type

I'll switch to decimal I guess :shrug: but - this is the generic jdbc generator, so maybe shouldn't index _too_ hard on redshift-specific stuff

**Review by gisripa — APPROVED** [review]

File: airbyte-cdk/java/airbyte-cdk/db-destinations/src/main/java/io/airbyte/cdk/integrations/destination/jdbc/typing_deduping/JdbcSqlGenerator.java [hunk 2]
```diff
@@ -62,8 +62,10 @@ protected DataType<?> toDialectType(final AirbyteType type) {
 
   protected DataType<?> toDialectType(final AirbyteProtocolType airbyteProtocolType) {
     return switch (airbyteProtocolType) {
-      case STRING -> SQLDataType.VARCHAR;
-      case NUMBER -> SQLDataType.FLOAT;
+      // Many destinations default to a very short length (e.g. Redshift defaults to 256)
+      case STRING -> SQLDataType.VARCHAR(65535);
+      // We default to precision=39, scale=9 across destinations
+      case NUMBER -> SQLDataType.DECIMAL(38, 9);
```
Comment (evantahler) [inline airbyte-cdk/java/airbyte-cdk/db-destinations/src/main/java/io/airbyte/cdk/integrations/destination/jdbc/typing_deduping/JdbcSqlGenerator.java:68]: 👍 and good comments

**Review by evantahler — APPROVED** [review]

File: airbyte-cdk/java/airbyte-cdk/db-destinations/src/main/java/io/airbyte/cdk/integrations/destination/jdbc/typing_deduping/JdbcSqlGenerator.java (hunk 2 above)
Comment (gisripa) [reply airbyte-cdk/java/airbyte-cdk/db-destinations/src/main/java/io/airbyte/cdk/integrations/destination/jdbc/typing_deduping/JdbcSqlGenerator.java:68]: 💯

Comment (evantahler) [reply airbyte-cdk/java/airbyte-cdk/db-destinations/src/main/java/io/airbyte/cdk/integrations/destination/jdbc/typing_deduping/JdbcSqlGenerator.java:68]: Nit/Note - these changes do appear to be for a generic JDBC driver not redshift itself... we might want to seperate these later.

File: airbyte-cdk/java/airbyte-cdk/db-destinations/src/main/java/io/airbyte/cdk/integrations/destination/jdbc/typing_deduping/JdbcSqlGenerator.java [hunk 3]
```diff
@@ -62,8 +62,10 @@ protected DataType<?> toDialectType(final AirbyteType type) {
 
   protected DataType<?> toDialectType(final AirbyteProtocolType airbyteProtocolType) {
     return switch (airbyteProtocolType) {
-      case STRING -> SQLDataType.VARCHAR;
-      case NUMBER -> SQLDataType.FLOAT;
+      // Many destinations default to a very short length (e.g. Redshift defaults to 256)
+      case STRING -> SQLDataType.VARCHAR(65535);
+      // We default to precision=39, scale=9 across destinations
```
Comment (cynthiaxyin) [inline airbyte-cdk/java/airbyte-cdk/db-destinations/src/main/java/io/airbyte/cdk/integrations/destination/jdbc/typing_deduping/JdbcSqlGenerator.java:67]: Is this 39 supposed to be 38? Also where does the 9 come from?

Comment (edgao) [reply airbyte-cdk/java/airbyte-cdk/db-destinations/src/main/java/io/airbyte/cdk/integrations/destination/jdbc/typing_deduping/JdbcSqlGenerator.java:67]: > 39

.... thanks copilot 😅 

> 9

mostly arbitrary. I think this is what redshift defaults to if you don't specify the parameters, and [bigquery](https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types#decimal_types) uses the same settings.

**Pull request merged**